File: Mira/Services/AccountService.swift
import Foundation
import AuthenticationServices
import Combine

enum MiraAuthError: LocalizedError {
    case deviceAlreadyRegistered
    var errorDescription: String? {
        "A free account already exists on this Mac. Sign in to that account, or upgrade to Pro to use Mira on multiple devices."
    }
}

// MARK: - User

struct MiraUser: Codable, Equatable {
    let id: String
    let email: String?
    let displayName: String?
    let avatarURL: String?
    let createdAt: Date
}

// MARK: - Auth State

enum AuthState: Equatable {
    case signedOut
    case signedIn(MiraUser)
    case loading
}

// MARK: - Service

@MainActor
final class AccountService: NSObject, ObservableObject {
    static let shared = AccountService()

    @Published private(set) var authState: AuthState = .signedOut

    private let userKey = "mira_current_user"

    override init() {
        super.init()
        // Prefer Supabase session if available
        if let s = SupabaseService.shared.session {
            let user = MiraUser(
                id: s.userId,
                email: s.email,
                displayName: s.displayName,
                avatarURL: nil,
                createdAt: Date()
            )
            authState = .signedIn(user)
            Task { await EntitlementService.shared.fetchAndApplyPlan() }
        } else {
            loadSavedUser()
        }
    }

    var currentUser: MiraUser? {
        if case .signedIn(let user) = authState { return user }
        return nil
    }

    var isSignedIn: Bool { currentUser != nil }

    // MARK: - Apple Sign In

    func handleAppleAuthorization(_ authorization: ASAuthorization) {
        guard let cred = authorization.credential as? ASAuthorizationAppleIDCredential else { return }
        let name = [cred.fullName?.givenName, cred.fullName?.familyName]
            .compactMap { $0 }.joined(separator: " ")
        Task { @MainActor in
            guard let tokenData = cred.identityToken,
                  let idToken = String(data: tokenData, encoding: .utf8) else {
                authState = .signedOut
                return
            }
            authState = .loading
            do {
                let s = try await SupabaseService.shared.signInWithApple(idToken: idToken)
                let user = MiraUser(
                    id: s.userId,
                    email: s.email ?? cred.email,
                    displayName: s.displayName ?? (name.isEmpty ? cred.email?.components(separatedBy: "@").first : name),
                    avatarURL: nil,
                    createdAt: Date()
                )
                saveUser(user)
                authState = .signedIn(user)
                PostHogService.shared.identify(userId: s.userId, email: s.email, name: s.displayName)
                await EntitlementService.shared.fetchAndApplyPlan()
                await registerDevice(DeviceFingerprintService.deviceHash, jwt: s.accessToken)
            } catch {
                authState = .signedOut
            }
        }
    }

    // MARK: - Email/Password via Supabase

    func signIn(email: String, password: String) async throws {
        authState = .loading
        do {
            let s = try await SupabaseService.shared.signIn(email: email, password: password)
            let user = MiraUser(
                id: s.userId,
                email: s.email,
                displayName: s.displayName ?? s.email?.components(separatedBy: "@").first,
                avatarURL: nil,
                createdAt: Date()
            )
            saveUser(user)
            authState = .signedIn(user)
            PostHogService.shared.identify(userId: s.userId, email: s.email, name: s.displayName)
            await EntitlementService.shared.fetchAndApplyPlan()
            Task { await registerDevice(DeviceFingerprintService.deviceHash, jwt: s.accessToken) }
        } catch {
            authState = .signedOut
            throw error
        }
    }

    func signUp(email: String, password: String, name: String) async throws {
        authState = .loading
        do {
            // Block signup if a free account already exists on this device
            let deviceHash = DeviceFingerprintService.deviceHash
            let available = await checkDeviceAvailable(deviceHash)
            guard available else {
                authState = .signedOut
                throw MiraAuthError.deviceAlreadyRegistered
            }

            guard let s = try await SupabaseService.shared.signUp(email: email, password: password, name: name) else {
                // Email confirmation required — not an error, just stay signed out.
                authState = .signedOut
                return
            }
            let user = MiraUser(
                id: s.userId,
                email: s.email,
                displayName: s.displayName ?? name,
                avatarURL: nil,
                createdAt: Date()
            )
            saveUser(user)
            authState = .signedIn(user)
            PostHogService.shared.identify(userId: s.userId, email: s.email, name: s.displayName)
            await EntitlementService.shared.fetchAndApplyPlan()
            Task { await registerDevice(deviceHash, jwt: s.accessToken) }
        } catch {
            authState = .signedOut
            throw error
        }
    }

    // MARK: - Device lock

    private func checkDeviceAvailable(_ hash: String) async -> Bool {
        guard let url = URL(string: AppSecrets.supabaseURL + "/functions/v1/check-device") else { return true }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue(AppSecrets.supabaseAnonKey, forHTTPHeaderField: "apikey")
        req.httpBody = try? JSONSerialization.data(withJSONObject: ["device_hash": hash])
        guard let (data, _) = try? await URLSession.shared.data(for: req),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let available = json["available"] as? Bool else { return true }
        return available
    }

    // Fire-and-forget after sign-in. If the server returns 409 (another free account
    // owns this device), we sign the user out so they can't use Mira on this Mac.
    private func registerDevice(_ hash: String, jwt: String) async {
        guard let url = URL(string: AppSecrets.supabaseURL + "/functions/v1/register-device") else { return }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("Bearer \(jwt)", forHTTPHeaderField: "Authorization")
        req.httpBody = try? JSONSerialization.data(withJSONObject: ["device_hash": hash])
        guard let (data, resp) = try? await URLSession.shared.data(for: req),
              let http = resp as? HTTPURLResponse else { return }
        if http.statusCode == 409 {
            let msg = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["error"] as? String ?? ""
            if msg == "device_already_registered" {
                await MainActor.run { signOut() }
            }
        }
    }

    func signOut() {
        SupabaseService.shared.signOut()
        authState = .signedOut
        UserDefaults.standard.removeObject(forKey: userKey)
    }

    // MARK: - Persistence

    private func saveUser(_ user: MiraUser) {
        if let data = try? JSONEncoder().encode(user) {
            UserDefaults.standard.set(data, forKey: userKey)
        }
    }

    private func loadSavedUser() {
        guard let data = UserDefaults.standard.data(forKey: userKey),
              let user = try? JSONDecoder().decode(MiraUser.self, from: data) else { return }
        authState = .signedIn(user)
    }
}

File: windows/Mira.Windows.Core/Account/AccountService.cs
using System.Net.Http.Headers;
using System.Text;
using Mira.Contracts.CheckDevice;
using Mira.Contracts.RegisterDevice;
using Mira.Windows.Core.Auth;
using Mira.Windows.Core.Config;
using Mira.Windows.Core.Device;
using Mira.Windows.Core.Entitlements;

namespace Mira.Windows.Core.Account;

/// <summary>
/// The Windows equivalent of macOS's <c>AccountService</c>
/// (Mira/Services/AccountService.swift) — orchestrates <see cref="SupabaseService"/>
/// (session), <see cref="DeviceFingerprintService"/> (device lock), and
/// <see cref="EntitlementService"/> (plan) into the single sign-in/sign-up flow a
/// UI actually calls. Deliberately narrower than the Swift original: email/password
/// only — no Sign in with Apple, no browser-OAuth fallback (both exist on macOS
/// only to route around an Apple-specific Developer-ID/App-Store entitlement
/// restriction that has no Windows analog; see
/// docs/windows/SECURITY_AND_PRIVACY.md §4). Adding Microsoft/Google OAuth later
/// is a product decision, not a porting gap.
/// </summary>
public sealed class AccountService
{
    private static readonly HttpClient Http = new();

    public static AccountService Shared { get; } = new();

    public AuthState State { get; private set; }
    public SupabaseSession? CurrentUser => SupabaseService.Shared.Session;
    public bool IsSignedIn => State == AuthState.SignedIn;

    public event Action<AuthState>? StateChanged;

    private AccountService()
    {
        State = SupabaseService.Shared.IsSignedIn ? AuthState.SignedIn : AuthState.SignedOut;
        if (State == AuthState.SignedIn) _ = EntitlementService.Shared.FetchAndApplyPlanAsync();
    }

    public async Task SignInAsync(string email, string password, CancellationToken ct = default)
    {
        SetState(AuthState.Loading);
        try
        {
            var session = await SupabaseService.Shared.SignInAsync(email, password, ct);
            SetState(AuthState.SignedIn);
            await EntitlementService.Shared.FetchAndApplyPlanAsync(ct);
            // Fire-and-forget, mirrors the Swift original's `Task { await registerDevice(...) }` —
            // a slow/failed device-registration call shouldn't block the sign-in UI.
            _ = RegisterDeviceAsync(session, ct);
        }
        catch
        {
            SetState(AuthState.SignedOut);
            throw;
        }
    }

    /// <summary>
    /// Throws <see cref="MiraAuthException"/> if this PC already hosts another
    /// free-tier account. Returns normally (with no session change) if signup
    /// succeeded but email confirmation is pending — mirrors the Swift original.
    /// </summary>
    public async Task SignUpAsync(string email, string password, string displayName, CancellationToken ct = default)
    {
        SetState(AuthState.Loading);
        try
        {
            var deviceHash = DeviceFingerprintService.DeviceHash;
            if (!await CheckDeviceAvailableAsync(deviceHash, ct))
            {
                SetState(AuthState.SignedOut);
                throw MiraAuthException.DeviceAlreadyRegistered();
            }

            var session = await SupabaseService.Shared.SignUpAsync(email, password, displayName, ct);
            if (session is null)
            {
                SetState(AuthState.SignedOut); // pending email confirmation — not an error
                return;
            }

            SetState(AuthState.SignedIn);
            await EntitlementService.Shared.FetchAndApplyPlanAsync(ct);
            _ = RegisterDeviceAsync(session, ct);
        }
        catch
        {
            SetState(AuthState.SignedOut);
            throw;
        }
    }

    public void SignOut()
    {
        SupabaseService.Shared.SignOut();
        SetState(AuthState.SignedOut);
    }

    // ---- Device lock ---------------------------------------------------------

    private static async Task<bool> CheckDeviceAvailableAsync(string deviceHash, CancellationToken ct)
    {
        // Public endpoint, called BEFORE a session exists — authorized with the
        // anon key only, mirroring AccountService.swift's checkDeviceAvailable().
        // Fails open (returns true = "available") on any transport/parse error,
        // exactly like the Swift original — a device-lock check that can't reach
        // the server must not block signup entirely.
        try
        {
            using var req = new HttpRequestMessage(HttpMethod.Post, MiraConfig.SupabaseUrl + "/functions/v1/check-device");
            req.Headers.Add("apikey", MiraConfig.SupabaseAnonKey);
            var body = new CheckDeviceRequest { DeviceHash = deviceHash };
            req.Content = new StringContent(body.ToJson(), Encoding.UTF8, "application/json");
            using var resp = await Http.SendAsync(req, ct);
            if (!resp.IsSuccessStatusCode) return true;
            var text = await resp.Content.ReadAsStringAsync(ct);
            return CheckDeviceResponse.FromJson(text).Available;
        }
        catch
        {
            return true;
        }
    }

    /// <summary>
    /// Fire-and-forget after sign-in/sign-up. If the server returns 409 (another
    /// free account owns this device), signs back out so this PC can't be used —
    /// mirrors AccountService.swift's registerDevice().
    /// </summary>
    private static async Task RegisterDeviceAsync(SupabaseSession session, CancellationToken ct)
    {
        try
        {
            var body = new RegisterDeviceRequest { DeviceHash = DeviceFingerprintService.DeviceHash };
            using var req = new HttpRequestMessage(HttpMethod.Post, MiraConfig.SupabaseUrl + "/functions/v1/register-device");
            req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", session.AccessToken);
            req.Content = new StringContent(body.ToJson(), Encoding.UTF8, "application/json");
            using var resp = await Http.SendAsync(req, ct);
            if (resp.StatusCode == System.Net.HttpStatusCode.Conflict)
            {
                SupabaseService.Shared.SignOut();
            }
        }
        catch
        {
            // Best-effort, mirrors the Swift original's `try?`-guarded fire-and-forget.
        }
    }

    private void SetState(AuthState state)
    {
        State = state;
        StateChanged?.Invoke(state);
    }
}
